### PR TITLE
AMQP-671: Upgrade to Log4j2-2.7 and others

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/AMQP-671

* Upgrade to Log4j2-2.7 and fix compatibility
* Upgrade to Spring Retry 1.2 and remove reflection manipulation in the `SimpleMessageListenerContainer`, when `StatefulRetryOperationsInterceptor` is detected to set `UseRawKey` to `true`
* Deprecate `MissingMessageIdAdvice`
* Upgrade to Jackson `2.8.4` and fix usage of deprecated API
* Upgrade to SF-4.3.4 since it is minimum in Spring Boot 1.5
* The `1.7` XSD
* `What's New` for `1.7`

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -114,7 +114,7 @@ subprojects { subproject ->
 	}
 
 	jacoco {
-		toolVersion = "0.7.2.201409121644"
+		toolVersion = "0.7.7.201606060606"
```
Comment (garyrussell) [inline build.gradle]: Is this compatible with Sonar? I believe we tried to update this some time ago and the sonar builds failed.

File: build.gradle [hunk 2]
```diff
@@ -242,7 +242,7 @@ project('spring-rabbit') {
 
 		compile ("org.apache.logging.log4j:log4j-core:$log4j2Version", optional)
 
-		testCompile "commons-cli:commons-cli:$commonsCliVersion"
+//		testCompile "commons-cli:commons-cli:$commonsCliVersion"
```
Comment (garyrussell) [inline build.gradle]: ??

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/retry/MissingMessageIdAdvice.java [hunk 3]
```diff
@@ -40,8 +40,11 @@
  * Must be placed before the retry interceptor in the advice chain.
  * @author Gary Russell
  * @since 1.1.2
+ * @deprecated since 1.7 in favor of non-retryable messages logic in the {@code StatefulRetryOperationsInterceptor}
  *
  */
+@Deprecated
+@SuppressWarnings("deprecation")
```
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/retry/MissingMessageIdAdvice.java]: We can't just deprecate it; it simply won't work any more with spring-retry 1.2.

File: build.gradle (hunk 1 above)
Comment (garyrussell) [inline build.gradle]: Thanks - the SI coverage on sonar looks ok, so it seems ok.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/retry/MissingMessageIdAdvice.java (hunk 3 above)
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/retry/MissingMessageIdAdvice.java]: I think we have no choice.

Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/retry/MissingMessageIdAdvice.java]: Or, revert the container changes and just unconditionally `setRawKey` to true.

File: build.gradle (hunk 2 above)
Comment (artembilan) [inline build.gradle]: I wondered why do we need this dependency...

And looks like there is nothing on the matter.

Although I have a problem with the `MessageListenerContainerRetryIntegrationTests.testStatefulRetryWithNoMessageIds()`, but not sure that it is related... I might have just missed something about Retry.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/retry/MissingMessageIdAdvice.java (hunk 3 above)
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/retry/MissingMessageIdAdvice.java]: Eh... As for me the last variant sounds better as a smooth movement from 1.6 to 2.0 over 1.7.

Although I admit that I don't understand this new Retry logic yet...

File: build.gradle (hunk 2 above)
Comment (artembilan) [inline build.gradle]: Well, there is no comments why we need that `commons-cli`: https://github.com/spring-projects/spring-amqp/commit/1150fe48997ab6f9a612d40a51b5c6051c5e7efe#diff-c197962302397baf3a4cc36463dce5eaR230.

Any ideas ?

Comment (garyrussell) [inline build.gradle]: I have no recollection as to why I added it - it's only a test dependency; let's just remove it, not just comment it out.

**Comment by garyrussell** [comment]
Merged as df261f24918bf76ba4e9e2b6ded8fb6e179aeb0e with minor what's new polish.

**Pull request closed without merging**